KubeVirt provider init

mfranczy opened the pull request:
**What this PR does / why we need it**:
KubeVirt provider which allows to create VMs on the OpenShift/Kubernetes clusters.
It doesn't bring any network or control plane configuration.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy|action
- target_group:   user|operator|developer
-->
```improvement operator
None
```

mfranczy committed:
kubevirt extension init

* add empty infrastructure controller,
* add empty controlplane controller and webhooks,
* first implementation of worker controller

Signed-off-by: Marcin Franczyk <marcin0franczyk@gmail.com>

stoyanr commented on charts/images.yaml:
```diff
@@ -0,0 +1,17 @@
+images:
+- name: kubevirt-cloud-controller-manager
+  sourceRepository: github.com/kubevirt/cloud-provider-kubevirt
+  repository: docker.io/mfranczy/kubevirt-cloud-controller-manager
```
Could we use an image from an "official" location. Alternatively, could we host this image somewhere under `eu.gcr.io/gardener-project/gardener` if we build it ourselves?

stoyanr replied:
Please fix this, you should now have permissions to push into our docker registry.

mfranczy replied:
done

stoyanr commented on charts/images.yaml:
```diff
@@ -0,0 +1,17 @@
+  tag: "v0.33.0"
+- name: machine-controller-manager-provider-kubevirt
+  sourceRepository: github.com/gardener/machine-controller-manager-provider-kubevirt
+  repository: docker.io/mfranczy/machine-controller-manager-provider-kubevirt
```
Could we host this image somewhere under eu.gcr.io/gardener-project/gardener?

mfranczy replied:
Yes, it would be the best, I asked about that in the past, however, I don't have enough permissions to do it and there was a suggestion to push it to my private repo just for now. If you could help me with permissions issue or give me instruction about the process how you push images to gcr then that would be great.

stoyanr replied:
I don't find this acceptable, even at this point. Let's try to give you permissions, or perhaps I could find a workaround.

stoyanr replied:
Please fix this, you should now have permissions to push into our docker registry.

mfranczy replied:
done

stoyanr commented on charts/internal/machine-controller-manager/seed/templates/deployment.yaml:
```diff
@@ -0,0 +1,110 @@
+        gardener.cloud/role: controlplane
+        app: kubernetes
+        role: machine-controller-manager
+        networking.gardener.cloud/all: allowed
```
```suggestion
        networking.gardener.cloud/to-dns: allowed
        networking.gardener.cloud/to-public-networks: allowed
        networking.gardener.cloud/to-private-networks: allowed
        networking.gardener.cloud/to-seed-apiserver: allowed
        networking.gardener.cloud/to-shoot-apiserver: allowed
        networking.gardener.cloud/from-prometheus: allowed
```

mfranczy replied:
done

stoyanr commented on charts/internal/machine-controller-manager/seed/templates/deployment.yaml:
```diff
@@ -0,0 +1,110 @@
+            cpu: 50m
+            memory: 64Mi
+          limits:
+            cpu: "3"
+            memory: 3000Mi
```
Why are these values so large? On other providers, I find significantly lower values:
```yaml
          limits:
            cpu: 350m
            memory: 256Mi
```
Does the MCM provider sidecar require all these resources, and why?

mfranczy replied:
no reason, that was a copy paste from vsphere, let me check this

mfranczy replied:
i changed limits to suggested. done.

stoyanr commented on charts/internal/machine-controller-manager/shoot/Chart.yaml:
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
```
Normally you would have in this chart also a role and a rolebinding to manipulate secrets in the shoot namespace. I don't see them here?

stoyanr commented on cmd/gardener-extension-provider-kubevirt/main.go:
```diff
@@ -0,0 +1,20 @@
+
+	"github.com/gardener/gardener-extension-provider-kubevirt/cmd/gardener-extension-provider-kubevirt/app"
+
+	"github.com/gardener/gardener-resource-manager/pkg/log"
```
```suggestion
	"github.com/gardener/gardener/extensions/pkg/log"
```

mfranczy replied:
done

stoyanr commented on cmd/gardener-extension-provider-kubevirt/main.go:
```diff
@@ -0,0 +1,20 @@
+package main
```
Please add copyright header everywhere where it's still missing. 
```suggestion
// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main
```

mfranczy replied:
done

stoyanr commented on cmd/gardener-extension-provider-kubevirt/main.go:
```diff
@@ -0,0 +1,20 @@
+	runtimelog.SetLogger(log.ZapLogger(false))
+	cmd := app.NewControllerManagerCommand(controller.SetupSignalHandlerContext())
+	if err := cmd.Execute(); err != nil {
+		controllercmd.LogErrAndExit(err, "error while starting KubeVirt extension")
```
```suggestion
		controllercmd.LogErrAndExit(err, "error executing the main controller command")
```

mfranczy replied:
done

stoyanr commented on cmd/gardener-extension-provider-kubevirt/app/app.go:
```diff
@@ -0,0 +1,181 @@
+import (
+	"context"
+	"fmt"
+	apiskubevirthelper "github.com/gardener/gardener-extension-provider-kubevirt/pkg/apis/kubevirt/helper"
```
Please order imports everywhere in 3 groups as follows:
1. Golang imports
2. Project imports (`github.com/gardener/gardener-extension-provider-kubevirt`)
3. All other imports
This is the convention used in gardener and all extensions.

mfranczy replied:
done

stoyanr replied:
I still see the above 2 imports in the wrong place?

stoyanr commented on cmd/gardener-extension-provider-kubevirt/app/app.go:
```diff
@@ -0,0 +1,181 @@
+		reconcileOpts = &controllercmd.ReconcilerOptions{}
+
+		// options for the health care controller
+		healthCareCtrlOpts = &controllercmd.ControllerOptions{
```
Please rename this to `healthCheckCtrlOpts`, since it's "healthcheck controller", not "healthcare controller"

mfranczy replied:
done

stoyanr commented on cmd/gardener-extension-provider-kubevirt/app/app.go:
```diff
@@ -0,0 +1,181 @@
+				}
+			}
+
+			// TODO: remove the call and function once the MachineClass CRD deployment is fixed in Gardener
+			if err := apiskubevirthelper.ApplyMachineClassCRDs(ctx, restOpts.Completed().Config); err != nil {
+				controllercmd.LogErrAndExit(err, "Error ensuring the machine class CRD")
+			}
+
```
Is this needed? I think it has been fixed meanwhile.

stoyanr commented on cmd/gardener-extension-provider-kubevirt/app/app.go:
```diff
@@ -0,0 +1,181 @@
+
+			// apply config options
+			configFileOpts.Completed().ApplyETCDStorage(&kubevirtcontrolplaneexposure.DefaultAddOptions.ETCDStorage)
+			configFileOpts.Completed().ApplyGardenId(&kubevirtcontrolplane.DefaultAddOptions.GardenId)
+			configFileOpts.Completed().ApplyGardenId(&kubevirtinfra.DefaultAddOptions.GardenId)
```
Is this needed? I don't find it in the code of other providers.

stoyanr commented on cmd/gardener-extension-provider-kubevirt/app/app.go:
```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ *
+ */
+
```
Please use a properly formatted copyright header:
```suggestion
// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
```

mfranczy replied:
done

stoyanr reviewed (changes requested):
Thanks for the PR! This is part one of the review. I am intentionally splitting it into several parts as the PR is quite large.

mfranczy commented on .github/ISSUE_TEMPLATE/bug.md:
done

mfranczy commented on .github/ISSUE_TEMPLATE/feature.md:
done

mfranczy commented on .github/pull_request_template.md:
done

mfranczy commented on Makefile:
```diff
@@ -0,0 +1,142 @@
+#		--webhook-config-server-host=0.0.0.0 \
+#		--webhook-config-server-port=9443 \
+#		--webhook-config-cert-dir=./example/validator-kubevirt-certs
+
```
done

mfranczy commented on Dockerfile:
```diff
@@ -0,0 +1,22 @@
+############# gardener-extension-validator-kubevirt
+FROM base AS gardener-extension-validator-kubevirt
+
+#COPY --from=builder /go/bin/gardener-extension-validator-kubevirt /gardener-extension-validator-kubevirt
+#ENTRYPOINT ["/gardener-extension-validator-kubevirt"]
```
done

stoyanr replied:
I still see these commented lines?

mfranczy commented on Makefile:
```diff
@@ -0,0 +1,142 @@
+#
+#.PHONY: test-clean
+#test-clean:
+#	@$(REPO_ROOT)/vendor/github.com/gardener/gardener/hack/test-cover-clean.sh
```
done

mfranczy commented on charts/gardener-extension-provider-kubevirt/doc.go:
```diff
@@ -0,0 +1,20 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ *
+ */
```
done

mfranczy commented on charts/gardener-extension-provider-kubevirt/values.yaml:
```diff
@@ -0,0 +1,73 @@
+    region: someregion
+    spec:
+      provider:
+        providerConfig: {}
```
done

mfranczy commented on charts/gardener-extension-provider-kubevirt/values.yaml:
```diff
@@ -0,0 +1,73 @@
+#   ...
+
+webhookConfig:
+  serverPort: 443
```
done

mfranczy commented on charts/gardener-extension-provider-kubevirt/templates/service.yaml:
```diff
@@ -0,0 +1,19 @@
+metadata:
+  name: {{ include "name" . }}
+  namespace: {{ .Release.Namespace }}
+  {{-  if .Values.ignoreResources }}
+  annotations:
+    resources.gardener.cloud/ignore: "true"
+  {{- end }}
+  labels:
```
done

stoyanr replied:
I stil see these 2 lines indented?

mfranczy commented on charts/gardener-extension-provider-kubevirt/templates/rbac.yaml:
```diff
@@ -0,0 +1,100 @@
+  - pods/log
+  - mutatingwebhookconfigurations
+  - customresourcedefinitions
+  - networkpolicies
+  - csidrivers
```
csidrivers right now are not needed, however, networkpolicies I must test, I thought maybe there is some requirement for the extension pkgs.

mfranczy commented on charts/gardener-extension-provider-kubevirt/templates/rbac.yaml:
```diff
@@ -0,0 +1,100 @@
+  - rbac.authorization.k8s.io
+  - admissionregistration.k8s.io
+  - apiextensions.k8s.io
+  - networking.k8s.io
+  - storage.k8s.io
```
same as https://github.com/gardener/gardener-extension-provider-kubevirt/pull/1#discussion_r465688313

mfranczy commented on charts/gardener-extension-provider-kubevirt/values.yaml:
```diff
@@ -0,0 +1,73 @@
+image:
+  repository: mfranczy/gardener-extension-provider-kubevirt
```
https://github.com/gardener/gardener-extension-provider-kubevirt/pull/1#discussion_r465679072

stoyanr replied:
Please fix this, you should now have permissions to push into our docker registry.

mfranczy replied:
done

mfranczy commented on charts/gardener-extension-provider-kubevirt/templates/deployment.yaml:
```diff
@@ -0,0 +1,93 @@
+  {{-  if .Values.ignoreResources }}
+  annotations:
+    resources.gardener.cloud/ignore: "true"
+  {{- end }}
```
done

mfranczy replied:
btw. this feature didn't work for me, I saw that controller was still reconciling some CRs, is this annotation being read by extension pkgs? Like the main reconciliation loops for infra, controlplane and worker CRs?

stoyanr replied:
What exactly are you trying to achieve? The annotation is meant for the so called "gardener resource manager" when it tries to reconcile "managed resources". The extension controller itself is managed in the seed as such a resource, so this annotation would prevent GRM from reconciling it. Is this really what you want, and why? I didn't comment much on this so far because I thought it could be addressed later, and I haven't had the chance to test the extension yet.

mfranczy replied:
I wanted to use it for the local setup, it means I would have to deploy the controller registration CR (which is required) with the annotation and later start the extension controller locally from my desktop (`make start`). I found some documentation that says with the annotation the controller would not pick infra, control plane, worker CRs.

I decided to remove that annotation for now as I need to understand better the entire flow you described, and when I have a ready solution I will open a new PR for local dev setup. Maybe the doc I found will required an update too (I have to find that doc as I don't remember its location), if yes then I will create a PR there too.

stoyanr replied:
I think the purpose of the annotation is not what you had understood, see above. If you want to avoid reconciles from the cluster pods, the easiest way to achieve is to scale the extension deployment to 0 by setting the `replicaCount` value in the `controllerregistration` resource to 0.

mfranczy commented on Makefile:
```diff
@@ -0,0 +1,142 @@
+.PHONY: docker-images
+docker-images:
+	@docker build -t $(IMAGE_PREFIX)/$(NAME):$(VERSION)           -t $(IMAGE_PREFIX)/$(NAME):latest           -f Dockerfile -m 6g --target $(EXTENSION_PREFIX)-$(NAME)           .
+#	@docker build -t $(IMAGE_PREFIX)/$(VALIDATOR_NAME):$(VERSION) -t $(IMAGE_PREFIX)/$(VALIDATOR_NAME):latest -f Dockerfile -m 6g --target $(EXTENSION_PREFIX)-$(VALIDATOR_NAME) .
```
done

stoyanr commented on pkg/webhook/controlplane/add.go:
```diff
@@ -0,0 +1,58 @@
+	logger = log.Log.WithName("kubevirt-controlplane-webhook")
+)
+
+// AddOptions are options to apply when adding the KubeVirt controlplane controller to the manager.
+type AddOptions struct {
+	// Controller are the controller.Options.
+	Controller controller.Options
+	// IgnoreOperationAnnotation specifies whether to ignore the operation annotation or not.
+	IgnoreOperationAnnotation bool
+	// GardenId is the Gardener garden identity
+	GardenId string
+}
```
I don't think these are used anywhere?
```suggestion
```

mfranczy replied:
removed

stoyanr commented on pkg/webhook/controlplaneexposure/add.go:
```diff
@@ -0,0 +1,60 @@
+package controlplaneexposure
+
+import (
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
```
Please fix imports ordering, see above.

stoyanr reviewed (changes requested):
In general:
* Please ensure license header is present in all files, uses `//`, and the year is 2020.
* Please ensure the recommended order of imports is used in all Go files.

mfranczy commented:
> In general:
> Please ensure license header is present in all files, uses //, and the year is 2020.
> Please ensure the recommended order of imports is used in all Go files.


It was already fixed in the previous commits

mfranczy commented on hook-me.sh:
```diff
@@ -0,0 +1,288 @@
+#!/usr/bin/env bash
```
the file has been removed already in previous comments

stoyanr reviewed (changes requested):
Removed some comments that were meanwhile fixed. Please also squash all commits (another Gardener policy is one commit per PR).

mfranczy commented on report.xml:
```diff
@@ -0,0 +1,36 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
not needed, removed

mfranczy commented on Makefile:
```diff
@@ -0,0 +1,142 @@
+EXTENSION_PREFIX            := gardener-extension
+NAME                        := provider-kubevirt
+VALIDATOR_NAME              := validator-kubevirt
+REGISTRY                    := eu.gcr.io/gardener-project
```
fixed

mfranczy commented on example/00-componentconfig.yaml:
```diff
@@ -0,0 +1,16 @@
+machineImages:
+  - name: ubuntu
+    version: "16.04"
+    sourceURL: "https://cloud-images.ubuntu.com/xenial/current/xenial-server-cloudimg-amd64-disk1.img"
```
fixed

mfranczy commented on pkg/apis/kubevirt/helper/scheme_test.go:
```diff
@@ -0,0 +1,107 @@
+
+			Expect(err).To(BeNil())
+			fmt.Println("RESNEW", result.MachineImages)
+			fmt.Println(err)
```
done

mfranczy commented on pkg/apis/kubevirt/v1alpha1/types_infrastructure.go:
```diff
@@ -0,0 +1,38 @@
+type InfrastructureStatus struct {
+	metav1.TypeMeta
+
+	CreationStarted *bool
```
the field is not needed, removed

mfranczy commented on pkg/apis/kubevirt/v1alpha1/types_infrastructure.go:
```diff
@@ -0,0 +1,38 @@
+
+// InfrastructureConfig infrastructure configuration resource
+type InfrastructureConfig struct {
+	metav1.TypeMeta
```
done

mfranczy commented on pkg/apis/kubevirt/v1alpha1/types_infrastructure.go:
```diff
@@ -0,0 +1,38 @@
+
+// InfrastructureStatus contains information about created infrastructure resources.
+type InfrastructureStatus struct {
+	metav1.TypeMeta
```
done

mfranczy commented on pkg/cmd/options.go:
```diff
@@ -0,0 +1,49 @@
+	extensioncontrolplanewebhook "github.com/gardener/gardener/extensions/pkg/webhook/controlplane"
+)
+
+// ControllerSwitchOptions are the controllercmd.SwitchOptions for the provider controller.
```
done

mfranczy commented on pkg/controller/worker/machines.go:
```diff
@@ -0,0 +1,174 @@
+	if secret.Data == nil {
+		return nil, fmt.Errorf("secret does not contain any data")
+	}
+
```
good point, fixed

mfranczy commented on pkg/webhook/controlplane/add.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+var (
+	DefaultAddOptions = AddOptions{}
```
you're right, removed

mfranczy commented:
> Removed some comments that were meanwhile fixed. Please also squash all commits (another Gardener policy is one commit per PR).

Done. I squashed the commits as you asked and addressed all comments. The last thing is about docker registry access.

rhdojun commented on README.md:
```diff
@@ -1,3 +1,57 @@
+
+You can run the controller locally on your machine by executing `make start`.
+
+Static code checks and tests can be executed by running `VERIFY=true make all`. We are using Go modules for Golang package dependency management and [Ginkgo](https://github.com/onsi/ginkgo)/[Gomega](https://github.com/onsi/gomega) for testing.
```
Not working for me.

```
16:40:41:gardener-extension-provider-kubevirt🍔  VERIFY=true make all
make: *** No rule to make target 'all'.  Stop.
```

stoyanr replied:
The correct command is `make verify`

rhdojun commented on README.md:
```diff
@@ -1,3 +1,57 @@
+
+## How to start using or developing this extension controller locally
+
+You can run the controller locally on your machine by executing `make start`.
```
Not working for me.

```
16:40:53:gardener-extension-provider-kubevirt🍔  make start
{"level":"error","ts":"2020-08-06T16:49:24.663+0200","msg":"Error ensuring the machine CRDs","error":"Get https://api.crc.testing:6443/api?timeout=32s: dial tcp 192.168.130.11:6443: connect: no route to host","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/user/coding/gardener-extension-provider-kubevirt/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/gardener/gardener/extensions/pkg/controller/cmd.LogErrAndExit\n\t/home/user/coding/gardener-extension-provider-kubevirt/vendor/github.com/gardener/gardener/extensions/pkg/controller/cmd/cmd.go:32\ngithub.com/gardener/gardener-extension-provider-kubevirt/cmd/gardener-extension-provider-kubevirt/app.NewControllerManagerCommand.func1\n\t/home/user/coding/gardener-extension-provider-kubevirt/cmd/gardener-extension-provider-kubevirt/app/app.go:116\ngithub.com/spf13/cobra.(*Command).execute\n\t/home/user/coding/gardener-extension-provider-kubevirt/vendor/github.com/spf13/cobra/command.go:844\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/home/user/coding/gardener-extension-provider-kubevirt/vendor/github.com/spf13/cobra/command.go:945\ngithub.com/spf13/cobra.(*Command).Execute\n\t/home/user/coding/gardener-extension-provider-kubevirt/vendor/github.com/spf13/cobra/command.go:885\nmain.main\n\t/home/user/coding/gardener-extension-provider-kubevirt/cmd/gardener-extension-provider-kubevirt/main.go:17\nruntime.main\n\t/usr/lib/golang/src/runtime/proc.go:203"}
exit status 1
make: *** [Makefile:42: start] Error 1
```

rhdojun reviewed (commented):
Thanks for the new content @mfranczy! I'm also trying to have a look, but maybe the README is missing some prep steps?

mfranczy commented:
hey @rhdojun, thanks for review! Indeed you're right I have to update README.md

stoyanr commented:
@mfranczy Thanks for addressing all requests. The PR should be ok to merge from my perspective now. Updates to the README can be done in a later PR IMO. `make start` in particular can be tricky sometimes due to the webhooks.

stoyanr commented:
/lgtm
